### PR TITLE
implement RhoSpec contract in Rholang

### PR DESCRIPTION
**Pull request by odeac**
## Overview
+ add clues in the assertions so we can differentiate between different failures
+ fix the names and URNs of the assert primitives
+ migrate rholang tests to RhoSpec
+ remove the tests for BasicWallet and MakePoS as they are not easy to migrate while being irrelevant because their API will change soon
+ rename TestSetUtil to TestUtil

### JIRA ticket:
https://rchain.atlassian.net/browse/RCHAIN-2292

### Notes
The RhoSpec contract is just a demo of the assert primitives for now because I don't know yet how to sign and deploy a contract the way TestSet is deployed now. Once I figure that out I will move the implementation to a new file.

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

### [Bors](https://bors.tech/) cheat-sheet:

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

### CODE REVIEW

**Comment by KentShikama** [comment]
>  Why are we discarding all the old API's design choices, without even discussing them?

The don't think there was much discussion for the old API. I wouldn't treat it as a sacred battle tested piece of code.

> That's an equivalent of re-weaving our safety net while we're working 5 floors above it.

Over the past couple months we've been slowly tightening our safety net. At this point, the comm layer, block proposing and merging, the estimator, and contract execution in rholang is pretty solid. Bonding and safety oracle are being tightened up. Casper contracts (including wallets, REV, proof of stake) are still WIP and in some areas still not yet completely defined. Perhaps it is a matter of style but I think we can be a bit more aggressive with casper contracts.

**Comment by ArturGajowy** [comment]
>  I wouldn't treat it as a sacred battle tested piece of code.

I'm far from that. I'm just sure it's more battle-tested than starting from scratch.

> been slowly tightening our safety net

And now we're seemingly on the course to losing it due to errors in a big-bang revamp of the net's foundations.

Who tests the tests, after all? How can we tell the tests still catch the bugs they used to after we've rewritten them?

**Comment by KentShikama** [comment]
> And now we're seemingly on the course to losing it due to errors in a big-bang revamp of the net's foundations.

To be clear these tests have nothing to do with the already solid "comm layer, block proposing and merging, the estimator, and contract execution in rholang".

**Comment by ArturGajowy** [comment]
Are you then saying all the casper tests utilizing the `TestSet` contract are garbage and can be discarded?

**Comment by KentShikama** [comment]
> Are you then saying all the casper tests utilizing the TestSet contract are garbage and can be discarded?

Besides Either, ListOps, and NonNegativeNumber - I don't see those changing much. Depending on the outcome of the wallet and validator incentive discussions a lot of contracts could go through some drastic change.

**Comment by ArturGajowy** [comment]
It's not about whether any of the production code changes or not. It's about whether our test code can still be believed to secure (against) those changes and make them controllable.

**Comment by rabbitonweb** [comment]
I don't understand why there is work in this direction? @odeac haven't we had a meeting regarding this? 
we agreed that the framework has issues but putting sth new will be a challange and we have other places in code we should focus on. why have we spent 1h talking, taking conculsions, if we just gonna ignore the output of the meeting?

I second @ArturGajowy, my main concerns are following:
1. reimplementing existing test on a new contract is risky, as we might get false-positivies
2. changing current TestSet to apply to main issues (mainly ones that are mentioned in https://rchain.atlassian.net/browse/RCHAIN-2292) seem to be doable. They will require effort but small one comparing to creating new framework and reimplmenting tests. 

I'm not argueing against the feature, but timing does not seem right? You might say "well, hell, but the work is done, what's the problem?". my issue is that the work is not completed and there will be another huge effort moving this framework onto the existing test suite. or do you want to keep both?

**Comment by ArturGajowy** [comment]
So you're saying that of
```
MakePoSTest.rho
BasicWalletTest.rho
MakeMintTest.rho
EitherTest.rho
NonNegativeNumberTest.rho
```
we can discard all but the last two?

**Comment by KentShikama** [comment]
I wouldn't just discard them but their value isn't that large given how we will change the corresponding contracts.

**Comment by ArturGajowy** [comment]
Still, the new framework lacks usage examples (not a single test was ported so far) and it's clear porting won't be straightforward due to conflicting design paradigms ('all tests must be isolated' vs 'tests reuse the previous state and can be sequenced by the user'). While clean-slate design has its merits, it only works after identifying the use cases and requirements - and verifying them by examples.

If this is to be useful, it needs a battle test of porting all the existing tests. Given what you said @KentShikama, I'm now comfortable if this means rewriting them largely. But this should be done before we deem the new framework ready. Otherwise the solution is incomplete and only increases maintenance burden.

**Comment by dckc** [comment]
@rabbitonweb writes:
> why have we spent 1h talking, taking conculsions, if we just gonna ignore the output of the meeting?

I'm interested to look at the record of the meeting. Pointer, please?

If there is no record, well, that answers your question.

**Comment by ArturGajowy** [comment]
@odeac please reinstate a proper commit log. This PR is not reviewable otherwise, just like any other PR having commits bigger than ` at most 200 lines of code (excluding tests)` (which we require for **PRs**, and then even more so for all their commits)

File: casper/src/test/rholang/MakeMintTest.rho [hunk 1]
```diff
@@ -1,126 +1,117 @@
...
       }
     }
+  } |
+
+  new RhoSpecCh in {
+    rl!(`rho:id:6wnujzcraztjfg941skrtbdkdgbko8nuaqihuhn15s66oz8ro5gwbb`, *RhoSpecCh) |
+    for(@(_, RhoSpec) <- RhoSpecCh) {
+      @RhoSpec!("testSuite", *setup,
+        [
+          ("Purses should be created with the given balance", *test1),
```
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/rholang/MakeMintTest.rho]: option 4: Add support for annonymous contracts :)

Joke aside, I'll fix the names (option 1+2)

File: casper/src/test/scala/coop/rchain/casper/genesis/contracts/TestUtil.scala [hunk 2]
```diff
@@ -9,33 +9,31 @@ import coop.rchain.catscontrib.TaskContrib._
...
 
-  val testSetDeploy: Deploy = {
+  val rhoSpecDeploy: Deploy = {
     val deployData = DeployData(
       user = ProtoUtil.stringToByteString(
         "4ae94eb0b2d7df529f7ae68863221d5adda402fc54303a3d90a8a7a279326828"
       ),
       timestamp = 1539808849271L,
-      term = TestSet.code,
+      term = RhoSpec.code,
```
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/scala/coop/rchain/casper/genesis/contracts/TestUtil.scala]: ok

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala [hunk 3]
```diff
@@ -30,98 +37,235 @@ object IsBoolean {
...
+
+case class RhoAssertTrue(testName: String, value: Boolean, clue: String) extends RhoTestAssertion
+case class RhoAssertEquals(testName: String, expected: Any, actual: Any, clue: String)
+    extends RhoTestAssertion
+
+case class TestResult(assertions: Map[String, List[RhoTestAssertion]], hasFinished: Boolean) {
+  def addAssertion(assertion: RhoTestAssertion): TestResult = {
+    val newAssertion =
+      (assertion.testName, assertion :: assertions.getOrElse(assertion.testName, List.empty))
+    TestResult(assertions.+(newAssertion), hasFinished)
```
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: of course!

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala [hunk 4]
```diff
@@ -30,98 +37,235 @@ object IsBoolean {
...
 
 private object TestResultCollector {
   def apply[F[_]: Sync]: F[TestResultCollector[F]] =
     Ref
-      .of(List.empty[RhoTestAssertion])
+      .of(TestResult(Map.empty, hasFinished = false))
       .map(new TestResultCollector(_))
 }
 
+object RhoLogger {
```
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: done

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala [hunk 5]
```diff
@@ -30,98 +37,235 @@ object IsBoolean {
...
     }
 }
+
 object IsComparison {
-  def unapply(p: Seq[ListParWithRandomAndPhlos]): Option[(String, Par, String, Par)] =
+  def unapply(
+      p: Par
+  ): Option[(Par, String, Par)] =
+    p.singleExpr().collect {
+      case Expr(ETupleBody(ETuple(List(expected, IsString(operator), actual), _, _))) =>
```
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: for part 2 I don't understand how one can write an assertion like that

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala [hunk 6]
```diff
@@ -30,98 +37,235 @@ object IsBoolean {
...
+        } ++ Seq(
+        SystemProcess.Definition[Task](
+          "rho:io:stdlog",
+          Runtime.byteName(26),
+          2,
+          26L,
+          RhoLogger.handleMessage[Task](_)
+        )
+      )
+    TestUtil.runtime(testResultCollectorService)
```
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: done

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: Sth along the lines of `assert(HasLocallyFree[Expr].locallyFree(expr, 0).isEmpty)`

File: casper/src/test/rholang/MakeMintTest.rho (hunk 1 above)
[the remark this answers is not in the document]
Comment (ArturGajowy) [reply casper/src/test/rholang/MakeMintTest.rho]: :D

File: casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (odeac) [reply casper/src/test/scala/coop/rchain/casper/helper/RhoSpec.scala]: for part 2 added tests that RhoSpec behaves correctly for comparisons var vs var, const vs var and var vs const

**Pull request merged**